### PR TITLE
✨ ERC1155B

### PR DESCRIPTION
**Pull request by transmissions11**
## Description

Closes https://github.com/Rari-Capital/solmate/issues/189

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `dapp snapshot`?
- [x] Ran `npm run lint`?
- [x] Ran `forge test`?
- [x] Ran `dapp test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

File: src/tokens/ERC1155B.sol [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        require(
+            to.code.length == 0
+                ? to != address(0)
+                : ERC1155TokenReceiver(to).onERC1155Received(msg.sender, from, id, amount, data) ==
+                    ERC1155TokenReceiver.onERC1155Received.selector,
+            "UNSAFE_RECIPIENT"
+        );
+    }
+
+    function safeBatchTransferFrom(
```
Comment (dmfxyz) [inline src/tokens/ERC1155B.sol:98]: 👋 I think this, `_batchMint`, and `_batchBurn` can all move the `unchecked` to the loop level

**Review by dmfxyz — COMMENTED** [review]
✨aesthetics✨

File: src/tokens/ERC1155B.sol [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) public virtual {
+        require(msg.sender == from || isApprovedForAll[from][msg.sender], "NOT_AUTHORIZED");
```
Comment (pcaversaccio) [inline src/tokens/ERC1155B.sol:78]: This comment applies to all `require` operations: why not use custom errors? I mean it decreases deployment costs for sure as well as runtime costs if the revert condition is met. Also, I think including the address of the emitter in all custom errors would make sense. This helps end users identify which contract was reverted with a failed transaction, which is especially useful for complex transactions involving multiple contracts:

```solidity
error SomeError(address emitter, ..., ..., ...);

emit SomeError(address(this), ..., ..., ...);
```

But probably you're already opinionated about custom errors 🤪

**Comment by joshieDo** [comment]
Is there a rationale for using `memory` instead of `calldata` for certain function parameters (eg. arrays) ? It does seem to give better gas  results.

**Comment by transmissions11** [comment]
yea memory is worse but more composable for internal use... kinda conflicted cuz anyone serious about performance will probably fork anyway

**Comment by transmissions11** [comment]
well for example in `_mint`, if you make the `data` arg calldata then u can't do this

image

**Comment by dmfxyz** [comment]
imo sacrificing composability in a library/inherited class is not worth the calldata savings. solmate is about gas efficiency, but not at the cost of non-optimizoors having insane headaches building off the contract.

**Comment by transmissions11** [comment]
here's a compromise i think is fair: keep the internal methods (_mint) `memory` and the public ones `calldata` as those are the least likely to be called internally

**Pull request merged**